### PR TITLE
Update contributor docs

### LINKED ISSUE #498: Update contributor documentation
**Is your feature request related to a problem? Please describe.**

The contributor documentation (`docs/source/developer/contributing` and `docs/source/developer/documentation`) are very out of date and are a bit thin. They're what we wrote to get the core team started but aren't suitable as a guide for new developers or contributors.

**Describe the solution you'd like**
Take the recently updated docs for all of this from `pyrealm` and port to this project.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR is to update the contributor/developer docs to make them up to date and fit for an external audience.

Todo:

- [x] Import similar and recently updated docs from `pyrealm`, which has a very similar contributor setup.
- [x] Add new docs to table of contents and add some basic tags on outdated or new content.
- [x] Fix build errors to allow docs to be review online.
- [x] Convert `pyrealm` docs for `virtual_ecosystem`
- [x] Review content of outdated files and delete or port to new files as required.
- [x] Fix up the table of contents to the revised content.

Fixes #498

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works
- [ ] Relevant documentation reviewed and updated

### CODE REVIEW

**Comment by davidorme** [comment]
@vgro I've moved the files in. Now we need to review and adapt.

File: docs/source/development/documentation/documentation.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+be kept up to date with the literature for the project.
+
+The three common use cases are shown below using a couple of reference tags
+(`campbell_introduction_2012` and `porporato_hydrologic_2003`) that are included
+in the current [reference library](../../bibliography.md).
+
+* Cite with date in parentheses (``{cite:t}`campbell_introduction_2012` ``): the model
+  implemented in {cite:p}`campbell_introduction_2012`.
+* Cite with reference(s) in parentheses
+  (``{cite:p}`campbell_introduction_2012,porporato_hydrologic_2003` ``): using the P
```
Comment (vgro) [inline docs/source/development/documentation/documentation.md:122]: I assume this is also a relevant reference for the VE

File: docs/source/development/contributing/code_qa_and_typing.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+needs to install or update all the hooks that are applied to changes within a commit.
+Usually the hooks only run on files changed by a particular `git commit` but using
+`pre-commit run --all-files` scans the entire codebase and is a commonly used check to
+make sure all is well.
+
+### The `pre-commit` configuration
+
+The file
+[.pre-commit-config.yaml](https://github.com/ImperialCollegeLondon/pyrealm/blob/develop/.pre-commit-config.yaml)
+contains the pre-commit hooks used by `pyrealm`. The configuration file contains links
```
Comment (vgro) [inline docs/source/development/contributing/code_qa_and_typing.md:40]: please check this link

File: docs/source/development/contributing/code_qa_and_typing.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+`check for merge conflicts`
+: Checks for remaning `git` merge conflict markers in code files.
+
+`debug statements (python)`
+: Checks for debugger imports and `breakpoint()` calls, which should not end up in
+released code.
+
+`pyupgrade`
+: Updates Python syntax to the current Python 3.10 syntax.
+
```
Comment (vgro) [inline docs/source/development/contributing/code_qa_and_typing.md:52]: should this be replaced with `ruff`?

File: docs/source/development/contributing/code_qa_and_typing.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+structures.
+
+## Supressing checking
+
+The `pre-commit` tools sometimes complain about things that we do not want to change.
+Almost all of the tools can be told to suppress checking, using comments with a set
+format to tell the tool what to do.
+
+This should not be done lightly: we are using these QA tools for a reason.
+
```
Comment (vgro) [inline docs/source/development/contributing/code_qa_and_typing.md:138]: Again, might need updating to `ruff`?

File: docs/source/development/contributing/code_testing.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+Some of the package docstrings contain `doctest` examples of code use. These examples
+are intended to provide simple examples of method or function use and generate an
+output: the `doctest` module is used to make sure that the code runs and gives the
+expected result.
+
+We have configured `pytest` to automatically also run `doctest`, but you can manually
+check the tests in files using, for example:
+
+```bash
+poetry run python -m doctest pyrealm/pmodel/pmodel.py
```
Comment (vgro) [inline docs/source/development/contributing/code_testing.md:51]: Are we currently using doctest anywhere?

File: docs/source/development/contributing/overview.md [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+   failures.
+
+1. Request reviews from other package developers using the Review section on the PR
+   page. A PR cannot be merged into `develop` until at least one approving review has
+   been added to the code. Reviews will often suggest changes to the code and you should
+   discuss those suggestions and implement them.
+
+   Hopefully, you will have talked to other developers during the process of writing the
+   PR and should have some ideas of who to ask for a review. If not, please request
+   [`davidorme`](https://github.com/davidorme) to review the PR and we can then work out
```
Comment (vgro) [inline docs/source/development/contributing/overview.md:90]: would you like to keep this?

File: docs/source/development/contributing/overview.md [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+```sh
+git clone https://github.com/ImperialCollegeLondon/pyrealm.git
+```
+
+You can now use `poetry` to install the package dependencies. This is not just the
+package requirements for end users of the package, but also a wider set of tools used in
+package development. `poetry` uses the
+[pyproject.toml](https://github.com/ImperialCollegeLondon/pyrealm/blob/develop/pyproject.toml)
+file to configure the dependencies that will be installed.
```
Comment (vgro) [inline docs/source/development/contributing/overview.md:150]: please check link

File: docs/source/development/contributing/overview.md [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+Development of the `pyrealm` package uses [`pre-commit`](https://pre-commit.com/). This
+is a python tool that runs a set of checks on `git` commits and stops the commit from
+completing when any of those checks fail. We use `pre-commit` to help catch a wide range
+of common issues and make sure that all code pushed to the GitHub repository meets some
+simple quality assurance checks and uses some common formatting standards.
+
+There is a detailed description of the `pre-commit` output and the  configured checks
+and update process on the [code quality assurance page](./code_qa_and_typing.md).
+Briefly, we use `pre-commit` to catch inconsistent formatting and variable typing and to
+run the widely-used `flake8` code checking suite.
```
Comment (vgro) [inline docs/source/development/contributing/overview.md:211]: Again, is this replaced with `ruff`?

File: docs/source/development/contributing/overview.md [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+We use GitHub Action workflows to update `pre-commit`, run code quality checks on pull
+requests, and to automate profiling and release publication. See the [GitHub Actions
+page](./github_actions.md) for details.
+
+### Package version releases
+
+We use trusted publishing from GitHub releases to release new versions of `pyrealm` to
+[PyPI](https://pypi.org/project/pyrealm/). Releases are also picked up and archived on
+[Zenodo](https://doi.org/10.5281/zenodo.8366847). See the [release process
+page](./release_process.md) for details.
```
Comment (vgro) [inline docs/source/development/contributing/overview.md:285]: ??

File: docs/source/development/contributing/release_process.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+
+# Package release process
+
+The package release process has four stages. The last two steps are automated:
+
+* Merge changes from the `develop` branch onto `main` that will form the new release
+  version.
+* Publish a release on GitHub - this is basically just a specific tagged commit on
+  `main` that has some associated release notes.
+* Publish the code on Zenodo.
```
Comment (vgro) [inline docs/source/development/contributing/release_process.md:23]: do we publish on Zenodo?

File: docs/source/development/contributing/release_process.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+   This is different from having an explicit release version number (e.g. `X.Y.Za1`) -
+   it is just a marker used on GitHub.
+
+   At this point, you can either save the draft or simply publish it. It is probably
+   good practice to save the draft and then have a discussion with the other developers
+   about whether to publish it.
+
+1. Once everyone is agreed **publish the release**.
+
+## Publish the package on Zenodo
```
Comment (vgro) [inline docs/source/development/contributing/release_process.md:110]: Please check

**Comment by vgro** [comment]
I changed all the references from `pyrealm` to `virtual_ecosystem` and highlighted a few issues where I wasn't sure if that was still up to date.

File: docs/source/development/documentation/documentation.md (hunk 1 above)
Comment (davidorme) [reply docs/source/development/documentation/documentation.md:125]: I've swapped the original refs for something alraedy in our docs - more or less at random, so these could be better!

File: docs/source/development/contributing/overview.md (hunk 6 above)
Comment (davidorme) [reply docs/source/development/contributing/overview.md:90]: For now, I think.

File: docs/source/development/contributing/overview.md (hunk 9 above)
Comment (davidorme) [reply docs/source/development/contributing/overview.md]: We don't do this on Virtual Ecosystem. Truth be told, I'm not sure why we do it on `pyrealm`.

File: docs/source/development/contributing/release_process.md (hunk 10 above)
Comment (davidorme) [reply docs/source/development/contributing/release_process.md:23]: No.

File: docs/source/development/contributing/code_testing.md (hunk 5 above)
Comment (davidorme) [reply docs/source/development/contributing/code_testing.md:58]: We are now 😉

**Comment by davidorme** [comment]
@TaranRallings and @jacobcook1995 

Vivienne and I have been over this and updated for VE - can you have a look over the Development section updates. I've built the branch:

https://virtual-ecosystem.readthedocs.io/en/498-update-contributor-documentation/

Ignore the "The design of the Virtual Ecosystem" section. That still needs work.

**Review by TaranRallings — APPROVED** [review]
Holy! This development section looks great. 

The only thing I noticed was that in `Docstring style for the Virtual Ecosystem package` there is a `flake8` reference that might now need updating to rufus.

**Review by jacobcook1995 — APPROVED** [review]
LGTM!

**Pull request merged**